nitro-protocol/contracts/VirtualMarginApp.sol

```diff
@@ -0,0 +1,146 @@
+ */
+contract VirtualMarginApp is IForceMoveApp {
+
+    struct MarginState {
+        uint256 leaderAmount;
+        uint256 followerAmount;
+        uint256 version;                    // Highest number is the valid state
+        INitroTypes.Signature leaderSig;    // signature on abi.encode(channelId,amount)
+        INitroTypes.Signature followerSig;  // signature on abi.encode(channelId,amount)
+    }
+
+    enum AllocationIndices {
+        Leader,     // Leader is the virtual-channel initiator
+        Follower    // Follower accepted to establish the link
+    }
+
+    /**
+     * @notice Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @dev Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @param fixedPart Fixed part of the state channel.
+     * @param proof Array of recovered variable parts which constitutes a support proof for the candidate.
+     * @param candidate Recovered variable part the proof was supplied for.
+     */
+    function requireStateSupported(
+        FixedPart calldata fixedPart,
+        RecoveredVariablePart[] calldata proof,
+        RecoveredVariablePart calldata candidate
+    ) external pure override {
+        // This channel has only 4 states which can be supported:
+        // 0 prefund
+        // 1 postfund
+        // 2 redemption
+        // 3 final
+
+        // states 0,1,3 can be supported via unanimous consensus:
+
+        if (proof.length == 0) {
+            require(
+                NitroUtils.getClaimedSignersNum(candidate.signedBy) ==
+                    fixedPart.participants.length,
+                '!unanimous; |proof|=0'
+            );
+            if (candidate.variablePart.turnNum == 0) return; // prefund
+            if (candidate.variablePart.turnNum == 1) return; // postfund
+            if (candidate.variablePart.turnNum == 3) {
+                // final (note: there is a core protocol escape hatch for this, too, so it could be removed)
+                require(candidate.variablePart.isFinal, '!final; turnNum=3 && |proof|=0');
+                return;
+            }
+            revert('bad candidate turnNum; |proof|=0');
+        }
+
+        // State 2 can be supported via a forced transition from state 1:
+        //
+        //      (2)_B     [redemption state signed by Bob, includes a voucher signed by Alice. The outcome may be updated in favour of Bob]
+        //      ^
+        //      |
+        //      (1)_AIB   [fully signed postfund]
+
+        if (proof.length == 1) {
+            requireProofOfUnanimousConsensusOnPostFund(proof[0], fixedPart.participants.length);
+            require(candidate.variablePart.turnNum == 2, 'bad candidate turnNum; |proof|=1');
+            require(
+                NitroUtils.isClaimedSignedBy(candidate.signedBy, 2),
+                'redemption not signed by Bob'
+            );
```

Comment:
This can be omitted. We need to check signatures on either `VariablePart` or `MarginState`.